### PR TITLE
homogenization of testsuite syntax

### PR DESCRIPTION
**Pull request by lefessan**
The syntax of the testsuite is not homogeneous, so it makes it difficult to use other tools than `autoconf` to work with it.

The goal of this PR is to make it more homogeneous, and test/dev some other tooling to simplify running the testsuite and fixing issues (while still using `autoconf` testsuite for mainstream use).

### CODE REVIEW

File: tests/testsuite.src/data_pointer.at [hunk 1]
```diff
@@ -44,18 +44,18 @@ AT_CHECK([test "$COB_HAS_64_BIT_POINTER" = "yes"], [0], [], [],
 
 # Previous test "failed" --> 32 bit
 
-[AT_CHECK([$COBCRUN_DIRECT ./prog], [0], [0x00000000
```
Comment (GitMensch) [inline tests/testsuite.src/data_pointer.at:47]: This is using a format we commonly don't use, a conditional when a test "passed"/"failed". The [] scope these parts, so they should be kept in..

File: tests/testsuite.src/listings.at [hunk 2]
```diff
@@ -2558,7 +2558,7 @@ SIZE  TYPE           LVL  NAME                           PICTURE
 0 errors in compilation group
 ])
 
-[AT_CHECK([diff prog17-32.lst prog.lis], [0], [], [])]
+AT_CHECK([diff prog17-32.lst prog.lis], [0], [], [])
```
Comment (GitMensch) [inline tests/testsuite.src/listings.at:2561]: same as above

File: tests/testsuite.src/run_file.at [hunk 3]
```diff
@@ -251,7 +251,7 @@ AT_DATA([prog.cob], [
 
 AT_CHECK([$COMPILE prog.cob], [0], [], [])
 AT_CHECK([$COBCRUN_DIRECT ./prog], [0], [], [])
-[echo "Test" > ./nosubhere]
+echo "Test" > ./nosubhere
```
Comment (GitMensch) [inline tests/testsuite.src/run_file.at:254]: We may use `AT_DATA` here, what do you think?

File: tests/testsuite.src/run_file.at [hunk 4]
```diff
@@ -2880,10 +2880,10 @@ rewriting key: 98651234
 
 AT_CHECK([test "$COB_HAS_ISAM" = "db"], [0], [], [],
 # Previous test "failed" --> other ISAM, no need for more checks
-[AT_CHECK([true], [0])],
+AT_CHECK([true], [0]),
```
Comment (GitMensch) [inline tests/testsuite.src/run_file.at:2883]: same as above

File: tests/testsuite.src/run_fundamental.at [hunk 5]
```diff
@@ -826,7 +826,7 @@ AT_DATA([prog.cob], [
            ADD  1.0       TO DST2.
            IF DST1 not = DST2 THEN
                DISPLAY 'move/compare of very near numbers failed (not id
-      -                'entical) [added]: ' DST1 " - " DST2
+      -                'entical) [[added]]: ' DST1 " - " DST2
```
Comment (GitMensch) [inline tests/testsuite.src/run_fundamental.at:829]: I've reworded that

File: tests/testsuite.src/run_file.at [hunk 6]
```diff
@@ -695,7 +695,7 @@ AT_DATA([prog.cob], [
 ])
 
 AT_CHECK([$COMPILE -fassign-clause=external prog.cob], [0], [], [])
-AT_CAPTURE_FILE(./whatever)
+AT_CAPTURE_FILE([./whatever])
```
Comment (GitMensch) [inline tests/testsuite.src/run_file.at:698]: Those captures are a good catch. I've adjusted them (no need for the dot either):
AT_CAPTURE_FILE(./something) ->AT_CAPTURE_FILE([something])
AT_CAPTURE_FILE(SOME) -> AT_CAPTURE_FILE([SOME])

File: tests/testsuite.src/run_manual_screen.at [hunk 7]
```diff
@@ -24,7 +24,7 @@
 AT_SETUP([LINE])
 AT_KEYWORDS([screen])
 
-AT_SKIP_IF(test "$COB_HAS_CURSES" != "yes")
+AT_SKIP_IF([test "$COB_HAS_CURSES" != "yes"])
```
Comment (GitMensch) [inline tests/testsuite.src/run_manual_screen.at:27]: good catch, I've did the same

File: tests/testsuite.src/run_misc.at [hunk 8]
```diff
@@ -2819,7 +2819,7 @@ AT_CHECK([$COBCRUN_DIRECT ./prog], [0], [], [])
 AT_CLEANUP
 
 
-AT_SETUP([STRING / UNSTRING [NOT] ON OVERFLOW])
+AT_SETUP([STRING / UNSTRING [[NOT]] ON OVERFLOW])
```
Comment (GitMensch) [inline tests/testsuite.src/run_misc.at:2822]: taken

File: tests/testsuite.src/syn_file.at [hunk 9]
```diff
@@ -1128,7 +1128,7 @@ AT_DATA([prog.cob], [
        01  i-rec PIC X(10).
 ])
 
-/* FIXME: this test will false-positive fail on EBCDIC machines */
+# FIXME: this test will false-positive fail on EBCDIC machines
```
Comment (GitMensch) [inline tests/testsuite.src/syn_file.at:1131]: interesting that this ever worked, thanks for catching it

File: tests/testsuite.src/syn_misc.at [hunk 10]
```diff
@@ -1112,22 +1112,22 @@ AT_DATA([prog.cob], [
...
 
 # Previous test "failed" --> 32 bit
-[AT_CHECK([$COMPILE_ONLY prog.cob], [1], [],
+AT_CHECK([$COMPILE_ONLY prog.cob], [1], [],
 [prog.cob:6: error: 'SINGLE-ITEM' cannot be larger than 268435456 bytes
 prog.cob:8: error: 'FILLER 1' cannot be larger than 268435456 bytes
 prog.cob:7: error: 'GROUP-ITEM1' cannot be larger than 268435456 bytes
 prog.cob:9: error: 'GROUP-ITEM2' cannot be larger than 268435456 bytes
-])]
+])
```
Comment (GitMensch) [inline tests/testsuite.src/syn_misc.at:1120]: same as above

File: tests/testsuite.src/syn_multiply.at [hunk 11]
```diff
@@ -47,7 +47,7 @@ AT_DATA([prog.cob], [
            STOP RUN.
 ])
 
-AT_CHECK([$COMPILE_ONLY prog.cob], [1], ,
+AT_CHECK([$COMPILE_ONLY prog.cob], [1], [],
```
Comment (GitMensch) [inline tests/testsuite.src/syn_multiply.at:50]: already did that :-)

File: tests/testsuite.src/used_binaries.at [hunk 12]
```diff
@@ -323,19 +323,19 @@ AT_CHECK([test -f prog.s], [0], [], [],
 # Previous test "failed" --> prog.s not available --> likely a VS build
 # only check for file as cl.exe cannot create executables from self-created
 # assembler sources
-[AT_CHECK([test -f prog.asm], [0], [], [])],
+AT_CHECK([test -f prog.asm], [0], [], []),
```
Comment (GitMensch) [inline tests/testsuite.src/used_binaries.at:326]: same as above

**Review by GitMensch — COMMENTED** [review]
Thank you for taking the time to review the testsuite syntax.
I'm currently working on testsuite adjustments (because of adjusted/dropped msgids) and did part of that, but missed a lot of the others, which I've now included (the commit likely will need at least one day as I'm not finished with the msgid check).

Open points:
* I'm not sure yet about possibly using `AT_DATA` instead of `echo` - that would be reasonable, no?
* the seldom used format  "do something on failure/success of the previous test" likely should keep []

Just FYI:
* https://www.gnu.org/software/autoconf/manual/autoconf-2.71/html_node/Using-Autotest.html#Using-Autotest with included notes on writing and using the testsuite
* https://sourceforge.net/p/gnucobol/contrib/HEAD/tree/trunk/at_test_win/ for a 2017 approach to run the testsuite directly from its source via Java (which did work quite well back then)
* https://sourceforge.net/p/gnucobol/contrib/HEAD/tree/trunk/tools/gctest/ this is what Ron commonly uses to add new testcases (it may be useful to check if it outputs some "non homogenious parts")

**Comment by GitMensch** [comment]
Because of other pending commits I've checked in all of the testsuite changes that were related and done, please update the mirror and branch, enabling us to see what's "missing"; do you have any thoughts on the open points outlined above?

**Comment by lefessan** [comment]
I will update this PR on top of your commits, but here is a few points:

* I have a small prototype of tool that is able to run a large part of the tests, it's currently very basic, but I hope it will soon have some nice "modern" features. Given my agenda, I am only working on it from time to time, but it is on my top list when I have time for GnuCOBOL, as I am struggling more and more with the testsuite;

* About the current syntax: if we want tools to be able to use the test files, we should use as much M4 macros as possible, instead of using shell commands (that are extremely hard to parse...).

   * So, it is always better to use `AT_DATA` instead of `echo` (my tool for example can only handle single-line `echo` )
   * It would probably be better not to use `[ ... ]` to scope macros inside *run-if-fail* or *run-if-pass*,  because `[...]` is usually used to store *unparsed* string. Yet, it is still possible to recall the parser on these strings later, so it is probably still possible to make that work.
   * I don't know yet what to do with `if`s, but there are not so many of them...

**Comment by GitMensch** [comment]
Most of the conditional "if failing, then ..." are about 32/64 bit, you may come up with an
GC_TEST32
GC_TEST64
macro (ideally defined in testsuite.at), the ABI that does not match would then be ignored, the other one executed. This should reduce the issues and create not so many conflicts (writing and using more macros needs to be postponed until everything is merged to trunk).

**Comment by GitMensch** [comment]
Hm, I'll have a look at your suggestion next week after the update and push.

**Comment by lefessan** [comment]
I still need also to review it, in particular the replacements of `if` by `AT_CHECK`.

**Comment by lefessan** [comment]
For information, the program I wrote is here: https://github.com/ocamlpro/autofonce
It's main features:
* running the testsuite in parallel with minimal output
* creating a test by capturing a command execution

It requires this PR to be merge to fully work on gnucobol (once the workflows are fixed)

**Comment by lefessan** [comment]
I have a new version of the tool that is much more tolerant to the syntax, I am going to submit an updated version of this PR with much fewer changes.

File: tests/testsuite.src/run_file.at [hunk 13]
```diff
@@ -251,7 +251,7 @@ AT_DATA([prog.cob], [
 
 AT_CHECK([$COMPILE prog.cob], [0], [], [])
 AT_CHECK([$COBCRUN_DIRECT ./prog], [0], [], [])
-[echo "Test" > ./nosubhere]
+AT_CHECK([echo "Test" > ./nosubhere])
```
Comment (GitMensch) [inline tests/testsuite.src/run_file.at:254]: Please use AT_DATA here and below

**Review by GitMensch — APPROVED** [review]
the current version - 3 files only - can go to SVN any time, please switch the 2 echo to AT_DATA though.

File: tests/testsuite.src/data_pointer.at (hunk 1 above)
Comment (lefessan) [reply tests/testsuite.src/data_pointer.at]: Now, these `[...]` scopes are mandatory for my tool, so I actually had to add them in a few places.

**Comment by GitMensch** [comment]
feel free to check in

**Comment by lefessan** [comment]
Merged !
